src/content/tutorials/en/intro-to-p5-strands.mdx

```diff
@@ -0,0 +1,744 @@
+let bloomShader;
+let originalImage;
+
+function bloomCallback() {
+  const preBlur = uniformTexture(() => originalImage);
+
+  getColor((input, canvasContent) => {
+    const blurred = getTexture(canvasContent, input.texCoord);
+    const original = getTexture(preBlur, input.texCoord);
+
+    const intensity = max(original, 0.2) * 8;
+    
+    const bloom = original + blurred * intensity;
+    return [bloom.rgb, 1];
+  });
+}
+
+
+async function setup() {
+	createCanvas(200, 200, WEBGL);
+	originalImage = createFramebuffer();
+  bloomShader = baseFilterShader().modify(bloomCallback);
+}
+
+function draw() {
+	// Draw the previous code to a framebuffer, so that we can store it before our blur is applied.
+	originalImage.begin();
+  background(0);
+  rotateY(millis()/1000);
+  fill(40, 120, 255);
+  box(80);
+	originalImage.end();
+
+	imageMode(CENTER);
+	image(originalImage, 0, 0);
+	
+	// changing this value affects the spread of the bloom
+	filter(BLUR, 20);
+  filter(bloomShader);
+}
+`} />
+
+In the variable `brightPass`, we have two magic numbers which could be uniforms instead, or feel free to adjust the values to find something which looks good. The `0.2` inside of `max` acts as *a kind* of threshold. Areas which are full black in the blurred image will not be affected by bloom, as they will ultimately be multiplying by 0. The `8` multiplies the overall strength of the effect.
+
+We only select the `.rgb` components of the bloom, as otherwise our alpha will be higher than 1 which gives unexpected results. Since we have selected only some components, we have an opportunity to try 'swizzling'. Swizzling is a feature of GLSL and other shader languages which lets us select whichever components of a vector to construct a new one.
+
+Any combination of `.rgba`, `.xyzw`, or `.stpq` (for texture coordinates) can be accessed, or set like this. Each of these sets are an alias for the others, they ultimately just select value `[0, 1, 2, 3]` in array terms. In other words writing `col.xyzw` is the same as writing `col.rgba`.
+
+Try changing it to `.grg` or any other combination, for some last minute colour changes. This will construct a new vector `[col.g, col.g, col.r, 1]`. For example `[col.yyy, 1]` produces a greyscale output, as the `rgb` values are all the same.
+
+<Callout title="Hint">
+Try 'swizzling' the return value and writing something like 
+`inputs.color = [col.ggr, 1]`. 
+</ Callout>
+
+## Review
+We have written 4 shaders and learned to manipulate vertex and fragment shaders in p5.strands.
+
+- **Basic Color Shader** - We started with a simple modification of the base color shader, learning how to access and modify the final color of an object.
+- **Instanced Particles** - We tried our hand at GPU instancing and rendered hundreds of particles.
+- **Fresnel Edge Highlighting** - We made a more advanced effect to make a glowing edge on 3D objects.
+- **Post-processing** - We used two filter shaders to tie the scene together.

```

Comment:
Since the readers might have _just now_ learned fragment vs vertex shaders, I think it would be used to indicate for each of these for which are fragment shaders, and which is the vertex shader